src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,139 @@
+  }
+
+  datalog::RelationDeclaration GetDeclarationOrFatal(
+      absl::string_view relation_name);
+
+ private:
+  // Generates a new mapping from relation names to relation declarations
+  // implemented as a flat_hash_map. It is used in the implementation of
+  // the constructor for DeclarationEnvironment. This is implemented
+  // as a nested class so that it can access the DeclarationEnvironment's
+  // private state during the constructor.
+  class RelationDeclarationEnvironmentVisitor
+      : public AuthLogicAstTraversingVisitor<
+            RelationDeclarationEnvironmentVisitor> {
+   public:
+    RelationDeclarationEnvironmentVisitor(DeclarationEnvironment& enclosing_env)
+        : enclosing_env_(enclosing_env) {}
+
+   private:
+    Unit Visit(const datalog::RelationDeclaration& rel_decl) {
+      enclosing_env_.AddDeclaration(rel_decl);
+      return Unit();
+    }
+    DeclarationEnvironment& enclosing_env_;
+  };
+
+  absl::flat_hash_map<std::string, datalog::RelationDeclaration> inner_map_;
+};
+
+// A map from the names of principal / predicate argument literals
+// (the constants wrapped in quotes) to typings.
+class TypeEnvironment {
+ public:
+  TypeEnvironment(DeclarationEnvironment decl_env, const Program& prog);
+
+  datalog::ArgumentType GetTypingOrFatal(absl::string_view argument_name);
+
+  const absl::flat_hash_map<std::string, datalog::ArgumentType>& inner_map()
+      const {
+    return inner_map_;
+  }
+
+ private:
+  void AddTyping(absl::string_view arg_name, datalog::ArgumentType arg_type);
+
+  // This visitor aids in the construction of the TypeEnvironment.
+  // It populates a mapping (flat_hash_map) of constant (literal)
+  // names to their types by looking at the relations to which the literals
+  // are applied and declarations of those relations (including the types
+  // of the parameters where they are applied). It takes as input
+  // a DeclarationEnvironment.
+  // This is a nested class because it needs to access the private
+  // fields of the TypeEnvironment in the constructor implementation.
+  class TypeEnvironmentGenerationVisitor
+      : public AuthLogicAstTraversingVisitor<TypeEnvironmentGenerationVisitor> {
+   public:
+    TypeEnvironmentGenerationVisitor(DeclarationEnvironment decl_env,
+                                     TypeEnvironment& enclosing_env)
+        : decl_env_(decl_env), enclosing_env_(enclosing_env) {}
+
+   private:
+    Unit PreVisit(const Principal& principal);
+    Unit Visit(const datalog::Predicate& pred);
+    DeclarationEnvironment decl_env_;
+    TypeEnvironment& enclosing_env_;
+  };
+
+  absl::flat_hash_map<std::string, datalog::ArgumentType> inner_map_;
+};
+
+// Generate a new version of the program transformed to include
+// the universe declarations, universe populating rules,
+// and extensions of the rules to include conditions that check
+// these universes.
+Program InsertUniverseRelations(const Program& prog);
```

Comment:
This should be called just after the AST is generated from parsing and before the lowing_ast_to_datalog pass. This isn't called yet because that should probably come with integration tests and be a separate PR.